File: slidingWindow/maxConsecutiveOnes/1004_MaxConsecutiveOnes3.java
class Solution {
    // O(n) time O(1) space
    public int longestOnes(int[] A, int K) {
        int maxLen = 0;
        int windowStart = 0;
        for (int windowEnd = 0; windowEnd < A.length; windowEnd++) {
            if (A[windowEnd] == 0) K--;
            while (K < 0) {
                if (A[windowStart] == 0) K++;
                windowStart++;
            }
            maxLen = Math.max(maxLen, windowEnd - windowStart + 1);
        }
        return maxLen;
    }

    // What if the input numbers come in one by one as an infinite stream? 
    // In other words, you can't store all numbers coming from the stream as it's too large to hold in memory. Could you solve it efficiently?

    public int longestOnesStream(int[] A, int K) {
        int maxLen = 0;
        Deque<Integer> queue = new ArrayDeque<>();
        int windowStart = 0;
        for (int windowEnd = 0; windowEnd < A.length; windowEnd++) {
            if (A[windowEnd] == 0) queue.offer(windowEnd);
            if (queue.size() > K) {
                windowStart = queue.poll() + 1;
            }
            maxLen = Math.max(maxLen, windowEnd - windowStart + 1);
        }
        return maxLen;
    }
}

File: slidingWindow/maxConsecutiveOnes/1004_MaxConsecutiveOnes3.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        start = 0
        end = 0
        num_zeroes = 0
        max_consecutive_ones = 0
        while end < len(nums):
            if nums[end] == 0:
                num_zeroes += 1

            while num_zeroes > k:
                if nums[start] == 0:
                    num_zeroes -= 1
                start += 1
            max_consecutive_ones = max(max_consecutive_ones, end - start + 1)
            end += 1
        return max_consecutive_ones

        